jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ByteChannelContentSource.java

```diff
@@ -0,0 +1,297 @@
+            }
+            else if (_buffer.isRetained())
+            {
+                _buffer.release();
+                _buffer = _byteBufferPool.acquire();
+            }
+
+            try
+            {
+                ByteBuffer byteBuffer = _buffer.getByteBuffer();
+                BufferUtil.clearToFill(byteBuffer);
+                if (_length >= 0)
+                    byteBuffer.limit((int)Math.min(_buffer.capacity(), _length - _totalRead));
+                int read = _byteChannel.read(byteBuffer);
+                BufferUtil.flipToFlush(byteBuffer, 0);
+                if (read == 0)
+                    return null;
+                if (read > 0)
+                {
+                    _totalRead += read;
+                    _buffer.retain();
+                    if (_length < 0 || _totalRead < _length)
+                        return Content.Chunk.asChunk(byteBuffer, false, _buffer);
+
+                    Content.Chunk last = Content.Chunk.asChunk(byteBuffer, true, _buffer);
+                    lockedSetTerminal(Content.Chunk.EOF);
+                    return last;
+                }
+                lockedSetTerminal(Content.Chunk.EOF);
+            }
+            catch (Throwable t)
+            {
+                lockedSetTerminal(Content.Chunk.from(t, true));
+            }
+        }
+        return _terminal;
+    }
+
+    @Override
+    public void fail(Throwable failure)
+    {
+        try (AutoLock ignored = lock.lock())
+        {
+            lockedSetTerminal(Content.Chunk.from(failure, true));
+        }
+    }
+
+    @Override
+    public long getLength()
+    {
+        return _length;
+    }
+
+    @Override
+    public boolean rewind()
+    {
+        try (AutoLock ignored = lock.lock())
+        {
+            // We can remove terminal condition for a rewind that is likely to occur
+            if (_terminal != null && !Content.Chunk.isFailure(_terminal) && (_byteChannel == null || _byteChannel instanceof SeekableByteChannel))
+                _terminal = null;
+
+            lockedEnsureOpenOrTerminal();
+            if (_terminal != null || _byteChannel == null || !_byteChannel.isOpen())
+                return false;
+
+            if (_offset >= 0 && _byteChannel instanceof SeekableByteChannel seekableByteChannel)
+            {
+                try
+                {
+                    seekableByteChannel.position(_offset);
+                    _totalRead = 0;
+                    return true;
+                }
+                catch (Throwable t)
+                {
+                    lockedSetTerminal(Content.Chunk.from(t, true));
+                }
+            }
+            return false;
+        }
+    }
+
+    /**
+     * A {@link ByteChannelContentSource} for a {@link Path}
+     * @deprecated To be replaced by an updated {@link org.eclipse.jetty.io.content.PathContentSource} in 12.1.0
+     */
+    @Deprecated(forRemoval = true, since = "12.0.11")
```

Comment:
Weird that this is already deprecated without a replacement available, and is still used in many places in the code.

Reply:
yeah.... well I've been pushed in both directions by other review.  Ultimately we don't want to change the behavior on the other `PathContentSource` changed in a dot release, but we want to establish the new class with better semantic now.  It is just on notice that it will move to the other location in 12.1.  What else can we do to introduce a new semantic but not change the old until a point release.